File: Examples/intel_examples/examples_core_f/vmlf/source/vzpowx.f
!===============================================================================
! Copyright 2001-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    vzPowx  Example Program Text
!*******************************************************************************

      include "_rms.fi"

      program MKL_VML_TEST

      include "mkl_vml.f90"
      real(kind=4) :: srelerr
      real(kind=8) :: drelerr
      real(kind=4) :: crelerr
      real(kind=8) :: zrelerr

      complex(kind=8) zA(10)
      complex(kind=8) zB(10)
      complex(kind=8) zBha0(10)
      complex(kind=8) zBha1(10)
      complex(kind=8) zBha2(10)
      complex(kind=8) zBla1(10)
      complex(kind=8) zBla2(10)
      complex(kind=8) zBep1(10)
      complex(kind=8) zBep2(10)

      real(kind=8) CurRMS,MaxRMS

      integer, parameter :: inca=3
      integer, parameter :: incb=5
      complex(kind=8) zA_I(10*inca)
      complex(kind=8) zB_I(10*incb)
      complex(kind=8) zBha0_I(10*incb)
      complex(kind=8) zBha1_I(10*incb)
      complex(kind=8) zBha2_I(10*incb)
      complex(kind=8) zBla1_I(10*incb)
      complex(kind=8) zBla2_I(10*incb)
      complex(kind=8) zBep1_I(10*incb)
      complex(kind=8) zBep2_I(10*incb)
      real(kind=8) CurRMS_I,MaxRMS_I

      integer(kind=8) mode
      integer tmode
      integer i, vec_len

      vec_len=10
      MaxRMS=0.0
      MaxRMS_I=0.0

      zA( 1)=(0.1000d+000,7.0000d+000)
      zA( 2)=(0.8666d+000,6.2333d+000)
      zA( 3)=(1.6333d+000,5.4666d+000)
      zA( 4)=(2.4000d+000,4.6999d+000)
      zA( 5)=(3.1666d+000,3.9333d+000)
      zA( 6)=(3.9333d+000,3.1666d+000)
      zA( 7)=(4.7000d+000,2.3999d+000)
      zA( 8)=(5.4666d+000,1.6333d+000)
      zA( 9)=(6.2333d+000,0.8666d+000)
      zA(10)=(7.0000d+000,0.0999d+000)
      zB( 1)=(1.4660080920771602d+001,-4.2457663156400498d+000)
      zB( 2)=(6.6546252673495720d+000,-1.3285048886353213d+001)
      zB( 3)=(-7.0387482202786016d+000,-1.4767324650988847d+001)
      zB( 4)=(-2.1117845706227051d+001,-3.4930900830543785d+000)
      zB( 5)=(-2.4424383742451884d+001,2.4394003554591237d+001)
      zB( 6)=(4.1042328112091866d+000,6.8100661769434140d+001)
      zB( 7)=(1.0377465385856878d+002,1.1602928377093086d+002)
      zB( 8)=(3.5030954960170573d+002,1.3713895026288438d+002)
      zB( 9)=(8.9261595917122804d+002,7.6109202498160059d+001)
      zB(10)=(2.0121481862211938d+003,-1.3059786621885627d+002)

      do i=1,10
          zA_I((i-1)*inca+1)=zA(i)
          zB_I((i-1)*incb+1)=zB(i)
      end do

      call VZPOWX(vec_len,zA,zA(6),zBha0)
      call VZPOWXI(vec_len,zA_I,inca,zA(6),zBha0_I,incb)

      mode=VML_EP
      call VMZPOWX(vec_len,zA,zA(6),zBep1,mode)
      call VMZPOWXI(vec_len,zA_I,inca,zA(6),zBep1_I,incb,mode)

      tmode=VML_EP
      tmode=VMLSETMODE(tmode)
      call VZPOWX(vec_len,zA,zA(6),zBep2)
      call VZPOWXI(vec_len,zA_I,inca,zA(6),zBep2_I,incb)

      mode=VML_LA
      call VMZPOWX(vec_len,zA,zA(6),zBla1,mode)
      call VMZPOWXI(vec_len,zA_I,inca,zA(6),zBla1_I,incb,mode)

      tmode=VML_LA
      tmode=VMLSETMODE(tmode)
      call VZPOWX(vec_len,zA,zA(6),zBla2)
      call VZPOWXI(vec_len,zA_I,inca,zA(6),zBla2_I,incb)

      mode=VML_HA
      call VMZPOWX(vec_len,zA,zA(6),zBha1,mode)
      call VMZPOWXI(vec_len,zA_I,inca,zA(6),zBha1_I,incb,mode)

      tmode=VML_HA
      tmode=VMLSETMODE(tmode)
      call VZPOWX(vec_len,zA,zA(6),zBha2)
      call VZPOWXI(vec_len,zA_I,inca,zA(6),zBha2_I,incb)

      do i=1,10
          if(zBha0(i) .ne. zBha1(i)) then
              print *,"Error! Difference between VZPOWX and"
              print *," VMZPOWX in VML_HA mode detected"
              stop 1
          endif
          if(zBha0_I((i-1)*incb+1) .ne. zBha1_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZPOWXI and"
              print *," VMZPOWXI in VML_HA mode detected"
              stop 1
          endif
          if(zBha1(i) .ne. zBha2(i)) then
              print *,"Error! Difference between VZPOWX and"
              print *," VMZPOWX in VML_HA mode detected"
              stop 1
          endif
          if(zBha1_I((i-1)*incb+1) .ne. zBha2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZPOWXI and"
              print *," VMZPOWXI in VML_HA mode detected"
              stop 1
          endif
          if(zBla1(i) .ne. zBla2(i)) then
              print *,"Error! Difference between VZPOWX and"
              print *," VMZPOWX in VML_LA mode detected"
              stop 1
          endif
          if(zBla1_I((i-1)*incb+1) .ne. zBla2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZPOWXI and"
              print *," VMZPOWXI in VML_LA mode detected"
              stop 1
          endif
          if(zBep1(i) .ne. zBep2(i)) then
              print *,"Error! Difference between VZPOWX and"
              print *," VMZPOWX in VML_EP mode detected"
              stop 1
          endif
          if(zBep1_I((i-1)*incb+1) .ne. zBep2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZPOWXI and"
              print *," VMZPOWXI in VML_EP mode detected"
              stop 1
          endif
      end do

      print *,"vzPowx test/example program"
      print *,""
      print *,"           Argument                     vzPowx"
      print *,"======================================================", &
     &        "========================"
      do i=1,vec_len
            print 10,zA(i),"    ",zBha0(i)
            CurRMS=zrelerr(zB(i),zBha0(i))
            if(CurRMS>MaxRMS) MaxRMS=CurRMS
      end do

      print *,""
      print *,"vzPowxI test/example program"
      print *,"           Argument                     vzPowx"
      print *,"======================================================", &
     &        "========================"
      do i=1,10
            print 10,zA_I((i-1)*inca+1),"    ",zBha0_I((i-1)*incb+1)
            CurRMS_I=zrelerr(zB_I((i-1)*incb+1),zBha0_I((i-1)*incb+1))
            if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I
      end do

      print *,""
      if(MaxRMS>=1e-5) then
            print 11,"Error! Relative accuracy is ",MaxRMS
            stop 1
      else
            print 11,"Relative accuracy is ",MaxRMS
      endif

      if(MaxRMS_I>=1e-5) then
            print 11,"Error! Relative strided accuracy is ",MaxRMS_I
            stop 1
      else
            print 11,"Relative strided accuracy is ",MaxRMS_I
      endif

10    format(E15.7,E15.7,A5,E15.7,E15.7)
11    format(A,F25.16)

      end

File: Examples/intel_examples/examples_core_c/vmlc/source/vzpowx.c
/*******************************************************************************
* Copyright 2001-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    vzPowx  Example Program Text
!******************************************************************************/

#include <stdio.h>
#include "mkl_vml.h"

#include "_rms.h"

#define INCA 3
#define INCB 5

int main()
{
  MKL_Complex16 zA[10],zB[10];
  MKL_Complex16 zBha0[10],zBha1[10],zBha2[10];
  MKL_Complex16           zBla1[10],zBla2[10];
  MKL_Complex16           zBep1[10],zBep2[10];
  float CurRMS,MaxRMS=0.0;

  MKL_Complex16 zA_I[10*INCA],zB_I[10*INCB];
  MKL_Complex16 zBha0_I[10*INCB],zBha1_I[10*INCB],zBha2_I[10*INCB];
  MKL_Complex16                  zBla1_I[10*INCB],zBla2_I[10*INCB];
  MKL_Complex16                  zBep1_I[10*INCB],zBep2_I[10*INCB];
  float CurRMS_I,MaxRMS_I=0.0;
  MKL_INT inca=INCA,incb=INCB;

  MKL_INT i=0,vec_len=10;

  zA[0].real=0.1000;zA[0].imag=7.0000;
  zA[1].real=0.8666;zA[1].imag=6.2333;
  zA[2].real=1.6333;zA[2].imag=5.4666;
  zA[3].real=2.4000;zA[3].imag=4.6999;
  zA[4].real=3.1666;zA[4].imag=3.9333;
  zA[5].real=3.9333;zA[5].imag=3.1666;
  zA[6].real=4.7000;zA[6].imag=2.3999;
  zA[7].real=5.4666;zA[7].imag=1.6333;
  zA[8].real=6.2333;zA[8].imag=0.8666;
  zA[9].real=7.0000;zA[9].imag=0.0999;
  zB[0].real=1.4660080920771602e+001;zB[0].imag=-4.2457663156400498e+000;
  zB[1].real=6.6546252673495720e+000;zB[1].imag=-1.3285048886353213e+001;
  zB[2].real=-7.0387482202786016e+000;zB[2].imag=-1.4767324650988847e+001;
  zB[3].real=-2.1117845706227051e+001;zB[3].imag=-3.4930900830543785e+000;
  zB[4].real=-2.4424383742451884e+001;zB[4].imag=2.4394003554591237e+001;
  zB[5].real=4.1042328112091866e+000;zB[5].imag=6.8100661769434140e+001;
  zB[6].real=1.0377465385856878e+002;zB[6].imag=1.1602928377093086e+002;
  zB[7].real=3.5030954960170573e+002;zB[7].imag=1.3713895026288438e+002;
  zB[8].real=8.9261595917122804e+002;zB[8].imag=7.6109202498160059e+001;
  zB[9].real=2.0121481862211938e+003;zB[9].imag=-1.3059786621885627e+002;

  for(i=0;i<10;i++) {
    zA_I[i*inca]=zA[i];
    zB_I[i*incb]=zB[i];
  }

  vzPowx(vec_len,zA,zA[5],zBha0);
  vzPowxI(vec_len,zA_I,inca,zA[5],zBha0_I,incb);

  vmzPowx(vec_len,zA,zA[5],zBep1,VML_EP);
  vmzPowxI(vec_len,zA_I,inca,zA[5],zBep1_I,incb,VML_EP);

  vmlSetMode(VML_EP);
  vzPowx(vec_len,zA,zA[5],zBep2);
  vzPowxI(vec_len,zA_I,inca,zA[5],zBep2_I,incb);

  vmzPowx(vec_len,zA,zA[5],zBla1,VML_LA);
  vmzPowxI(vec_len,zA_I,inca,zA[5],zBla1_I,incb,VML_LA);

  vmlSetMode(VML_LA);
  vzPowx(vec_len,zA,zA[5],zBla2);
  vzPowxI(vec_len,zA_I,inca,zA[5],zBla2_I,incb);

  vmzPowx(vec_len,zA,zA[5],zBha1,VML_HA);
  vmzPowxI(vec_len,zA_I,inca,zA[5],zBha1_I,incb,VML_HA);

  vmlSetMode(VML_HA);
  vzPowx(vec_len,zA,zA[5],zBha2);
  vzPowxI(vec_len,zA_I,inca,zA[5],zBha2_I,incb);

  for(i=0;i<10;i++) {
    if(zBha0[i].real!=zBha1[i].real || zBha0[i].imag!=zBha1[i].imag || zBha1[i].real!=zBha2[i].real ||
        zBha1[i].imag!=zBha2[i].imag) {
      printf("Error! Difference between vzPowx and vmzPowx in VML_HA mode detected.\n");
      return 1;
    }

    if(zBha0_I[i*incb].real!=zBha1_I[i*incb].real || zBha0_I[i*incb].imag!=zBha1_I[i*incb].imag ||
        zBha1_I[i*incb].real!=zBha2_I[i*incb].real || zBha1_I[i*incb].imag!=zBha2_I[i*incb].imag) {
      printf("Error! Difference between vzPowxI and vmzPowxI in VML_HA mode detected.\n");
      return 1;
    }

    if(zBla1[i].real!=zBla2[i].real || zBla1[i].imag!=zBla2[i].imag) {
      printf("Error! Difference between vzPowx and vmzPowx in VML_LA mode detected.\n");
      return 1;
    }

    if(zBla1_I[i*incb].real!=zBla2_I[i*incb].real || zBla1_I[i*incb].imag!=zBla2_I[i*incb].imag) {
      printf("Error! Difference between vzPowxI and vmzPowxI in VML_LA mode detected.\n");
      return 1;
    }

    if(zBep1[i].real!=zBep2[i].real || zBep1[i].imag!=zBep2[i].imag) {
      printf("Error! Difference between vzPowx and vmzPowx in VML_EP mode detected.\n");
      return 1;
    }

    if(zBep1_I[i*incb].real!=zBep2_I[i*incb].real || zBep1_I[i*incb].imag!=zBep2_I[i*incb].imag) {
      printf("Error! Difference between vzPowxI and vmzPowxI in VML_EP mode detected.\n");
      return 1;
    }
  }

  printf("vzPowx test/example program\n\n");
  printf("           Argument                           vzPowx\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .2f %+.2f*i   % .2f %+.2f*i      % .5f %+.5f*i\n",zA[i].real,zA[i].imag,zA[5].real,zA[5].imag,
        zBha0[i].real,zBha0[i].imag);
    CurRMS=zrelerr(zB[i],zBha0[i]);
    if(CurRMS>MaxRMS) MaxRMS=CurRMS;
  }
  printf("\n");
  printf("vzPowxI test/example program\n\n");
  printf("           Argument                           vzPowx\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .2f %+.2f*i   % .2f %+.2f*i      % .5f %+.5f*i\n",zA_I[i*inca].real,zA_I[i*inca].imag,zA[5].real,
        zA[5].imag,zBha0_I[i*incb].real,zBha0_I[i*incb].imag);
    CurRMS_I=zrelerr(zB_I[i*incb],zBha0_I[i*incb]);
    if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I;
  }
  printf("\n");
  if(MaxRMS>=1e-5) {
    printf("Error! Relative accuracy is %.16f\n",MaxRMS);
    return 1;
  }
  else {
    printf("Relative accuracy is %.16f\n",MaxRMS);
  }

  printf("\n");

  if(MaxRMS_I>=1e-5) {
    printf("Error! Relative strided accuracy is %.16f\n",MaxRMS_I);
    return 1;
  }
  else {
    printf("Relative strided accuracy is %.16f\n",MaxRMS_I);
  }
  return 0;
}
